feat(kumacp): add mesh access log docs (#1097)
kumahq/kuma-website: app/docs/dev/policies/meshaccesslog.md

| Command Operator                     | Description                                                      |
|--------------------------------------|------------------------------------------------------------------|
| `%KUMA_MESH%`                        | name of the mesh in which traffic is flowing                     |
| `%KUMA_SOURCE_SERVICE%`              | name of a `service` that is the `source` of traffic              |
| `%KUMA_DESTINATION_SERVICE%`         | name of a `service` that is the `destination` of traffic         |
| `%KUMA_SOURCE_ADDRESS_WITHOUT_PORT%` | address of a `Dataplane` that is the `source` of traffic         |
| `%KUMA_TRAFFIC_DIRECTION%`           | direction of the traffic, `INBOUND`, `OUTBOUND` or `UNSPECIFIED` |

All additional access log _command operators_ are valid to use with both `TCP` and `HTTP` traffic.

If a _command operator_ is specific to `HTTP` traffic, such as `%REQ(X?Y):Z%` or `%RESP(X?Y):Z%`, it will be replaced by a symbol "`-`" in case of `TCP` traffic.

> Internally, Kuma [determines traffic protocol](/docs/{{ page.version }}/policies/protocol-support-in-kuma) based on the value of `kuma.io/protocol` tag on the `inbound` interface of a `destination` `Dataplane`.

Author: In the old policy we had a section here:

```
% tip %

To provide different format for TCP and HTTP logging you can define two `MeshAccessLog` entities, one for TCP and one for HTTP with matching `kuma.io/protocol` selector.

% endtip %
```

But if I understand the matching matrix right this won't work yet because we don't support MeshServiceSubset. @michaelbeaumont - can you confirm?

Reviewer: Yeah right now I think `MeshServiceSubset` is rejected via validation so we can leave this out

Reviewer: Right now I think `MeshServiceSubset` is rejected via validation so we can leave this out

Reviewer 2: Yeah that's where we need some tags to be service specific not inbound specific :(

---
title: MeshAccessLog (beta)
---

With the MeshAccessLog policy you can easily set up access logs on every data plane in a mesh.

{% warning %}
This policy uses new policy matching algorithm and is in beta state,
it should not be mixed with [TrafficLog](/docs/{{ page.version }}/policies/traffic-log).
{% endwarning %}

{% tip %}
In the rest of this page we assume you have already configured your observability tools to work with Kuma.
If you haven't already read the [observability docs](/docs/{{ page.version }}/explore/observability).
{% endtip %}

## TargetRef support matrix

| TargetRef type    | top level | to  | from |
|-------------------|-----------|-----|------|
| Mesh              | ✅         | ✅   | ✅    |
| MeshSubset        | ✅         | ❌   | ❌    |
| MeshService       | ✅         | ✅   | ❌    |
| MeshServiceSubset | ✅         | ❌   | ❌    |

> If you don't understand this table you should read [matching docs](/docs/{{ page.version }}/policies/matching).

Reviewer 2: Feels like this could be a `{% %}` helper so that it's auto updated for future targetRef. Can we open an issue for this?

Author: https://github.com/kumahq/kuma-website/issues/1114

Internally, Kuma [determines traffic protocol](/docs/{{ page.version }}/policies/protocol-support-in-kuma) based on the value of `kuma.io/protocol` tag on the `inbound` interface of a `destination` `Dataplane`.

There are two types of `format` - `plain` and `json`.

Plain will accept a string with _command operators_ and produce a string output.

JSON will accept a list of key-value pairs that will produce a valid JSON object.

#### Plain

The default format string for `TCP` traffic is:

```
[%START_TIME%] %RESPONSE_FLAGS% %KUMA_MESH% %KUMA_SOURCE_ADDRESS_WITHOUT_PORT%(%KUMA_SOURCE_SERVICE%)->%UPSTREAM_HOST%(%KUMA_DESTINATION_SERVICE%) took %DURATION%ms, sent %BYTES_SENT% bytes, received: %BYTES_RECEIVED% bytes
```

The default format string for `HTTP` traffic is:

```
[%START_TIME%] %KUMA_MESH% "%REQ(:METHOD)% %REQ(X-ENVOY-ORIGINAL-PATH?:PATH)% %PROTOCOL%" %RESPONSE_CODE% %RESPONSE_FLAGS% %BYTES_RECEIVED% %BYTES_SENT% %DURATION% %RESP(X-ENVOY-UPSTREAM-SERVICE-TIME)% "%REQ(X-FORWARDED-FOR)%" "%REQ(USER-AGENT)%" "%REQ(X-REQUEST-ID)%" "%REQ(:AUTHORITY)%" "%KUMA_SOURCE_SERVICE%" "%KUMA_DESTINATION_SERVICE%" "%KUMA_SOURCE_ADDRESS_WITHOUT_PORT%" "%UPSTREAM_HOST%"

> ```

Reviewer 2: So how do we pick from these 2 types?

Author: A few paragraphs above we describe that we use `kuma.io/protocol` - https://github.com/kumahq/kuma-website/pull/1097/files#diff-fee054c5265ec14fa4800c086675e7fedc48eb57d8dc2a6e37b5f0005e61e485R58

Or do you mean between plain vs json?

Reviewer 2: I mean plain vs json

Author: I'm guessing this is user preference or the backend might actually require `JSON`. I'll add a sentence here.

    - key: "kuma_mesh"
      value: "%KUMA_MESH%"
    - key: 'method'
      value: '"%REQ(:METHOD)%'
    - key: "path"
      value: "%REQ(X-ENVOY-ORIGINAL-PATH?:PATH)%"
    - key: 'protocol'
      value: '%PROTOCOL%'
    - key: "response_code"
      value: "%RESPONSE_CODE%"
    - key: "response_flags"
      value: "%RESPONSE_FLAGS%"
    - key: "bytes_received"
      value: "%BYTES_RECEIVED%"
    - key: "bytes_sent"
      value: "%BYTES_SENT%"
    - key: "duration_ms"
      value: "%DURATION%"
    - key: "upstream_service_time"
      value: "%RESP(X-ENVOY-UPSTREAM-SERVICE-TIME)%"
    - key: 'x_forwarded_for'
      value: '"%REQ(X-FORWARDED-FOR)%"'
    - key: 'user_agent'
      value: '"%REQ(USER-AGENT)%"'
    - key: 'request_id'
      value: '"%REQ(X-REQUEST-ID)%"'
    - key: 'authority'
      value: '"%REQ(:AUTHORITY)%"'
    - key: "kuma_source_service"
      value: "%KUMA_SOURCE_SERVICE%"
    - key: "kuma_destination_service"
      value: "%KUMA_DESTINATION_SERVICE%"
    - key: "kuma_source_address_without_port"
      value: "%KUMA_SOURCE_ADDRESS_WITHOUT_PORT%"
    - key: "upstream_host"
      value: "%UPSTREAM_HOST%"
```
</div>

</details>

### Backends

#### TCP

> You can configure a TCP backend with an address:

Reviewer 2: Maybe explain what this will do

```yaml
backends:
  - tcp:
      address: 127.0.0.1:5000
```

#### File

You can configure a file backend with a path:

```yaml
backends:
  - file:
      path: /tmp/access.log
```

## Examples

{% tabs meshaccesslog useUrlFragment=false %}
{% tab meshaccesslog Kubernetes %}

Full example:
```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshAccessLog
metadata:
  name: default
  namespace: kuma-system
  labels:
    kuma.io/mesh: default # optional, defaults to `default` if unset
spec:
  targetRef:
    kind: Mesh

>   default:

Reviewer 2: These don't seem valid they don't have to and from

Reviewer 2: We should also probably explain the point of `to` and `from` and instead of having just Full examples and simple examples just have real examples.

.e.g: logging traffic coming from a single service on a specific dataplane, logging to multiple backends, log all incoming traffic, all outgoing traffic.

Author: To me a full example is something that shows all configuration options available. The best is to combine both of them, have some functional examples and a reference one showing all options (with optional ones).

    backends:
      - tcp:
          address: 127.0.0.1:5000
          format:
            json:
              - key: "start_time"
                value: "%START_TIME%"
      - file:
          path: /tmp/access.log
          format:
            plain: '[%START_TIME%]'

> ```

Reviewer 2: Worth explaining this will log to 2 different places

Apply the configuration with `kubectl apply -f [..]`.

{% endtab %}
{% tab meshaccesslog Universal %}

Full example:
```yaml
type: MeshAccessLog
name: default
mesh: default
spec:
  targetRef:
    kind: Mesh
  default:
    backends:
      - tcp:
          address: 127.0.0.1:5000
          format:
            json:
              - key: "start_time"
                value: "%START_TIME%"
      - file:
          path: /tmp/access.log
          format:
            plain: '[%START_TIME%]'
```

Apply the configuration with `kumactl apply -f [..]` or with the [HTTP API](../../reference/http-api).

{% endtab %}
{% endtabs %}

> ## Logging external services

Reviewer: Added below

> When running Kuma on Kubernetes you can also log the traffic to external services.

Reviewer 2: Why only on k8s?

Author: I'm guessing this is outdated docs from "Traffic Log". Will remove this and adjust in there as well.

To do it use `MeshService` as a `targetRef` target.

> In such case `%KUMA_DESTINATION_SERVICE%` will have value `external` and `%UPSTREAM_HOST%` will have an IP of the service.

Reviewer: ```suggestion
In such case `%KUMA_DESTINATION_SERVICE%` will have value `external` and `%UPSTREAM_HOST%` will have an IP of the service.

## Builtin gateway

You can select a builtin gateway using the `kuma.io/service` value. A current limitation is that traffic routed from a gateway to a service will be logged by that gateway as having destination `"*"`.
```

> In such case `%KUMA_DESTINATION_SERVICE%` will have value `external` and `%UPSTREAM_HOST%` will have an IP of the service.

Reviewer: To be clear here, this line is only true for traffic that _isn't_ via an `ExternalService`. Using an `ExternalService`, targeted with `MeshService`, gives the correct value of the `kuma.io/service` value.

Author: I think this section does not make sense in the new matching and is wrong in the old policy. 

> traffic that isn't via an ExternalService

does this mean it only happens with passthrough mode? we can't target that right? so it means it does not happen at all, right?

---
title: MeshAccessLog (beta)
---

> With the MeshAccessLog policy you can easily set up access logs on every data plane in a mesh.

Reviewer 3: ```suggestion
Using the MeshAccessLog policy, you can set up access logs on every data plane in a mesh.
```

{% warning %}

> This policy uses new policy matching algorithm and is in beta state,
> it should not be mixed with [TrafficLog](/docs/{{ page.version }}/policies/traffic-log).

Reviewer 3: ```suggestion
This policy uses a new policy matching algorithm and is in beta state. It should not be combined with [TrafficLog](/docs/{{ page.version }}/policies/traffic-log).
```

{% endwarning %}

{% tip %}

> In the rest of this page we assume you have already configured your observability tools to work with Kuma.
> If you haven't already read the [observability docs](/docs/{{ page.version }}/explore/observability).

Reviewer 3: ```suggestion
This guide assumes you have already configured your observability tools to work with Kuma.
If you haven't, see the [observability docs](/docs/{{ page.version }}/explore/observability).
```

{% endtip %}

## TargetRef support matrix

| TargetRef type    | top level | to  | from |
|-------------------|-----------|-----|------|
| Mesh              | ✅         | ✅   | ✅    |
| MeshSubset        | ✅         | ❌   | ❌    |
| MeshService       | ✅         | ✅   | ❌    |
| MeshServiceSubset | ✅         | ❌   | ❌    |

> If you don't understand this table you should read [matching docs](/docs/{{ page.version }}/policies/matching).

Reviewer 3: ```suggestion
To learn more about the information in this table, see the [matching docs](/docs/{{ page.version }}/policies/matching).
```

## Configuration

### Format

Kuma gives you full control over the format of the access logs.

The shape of a single log record is defined by a template string that uses [command operators](https://www.envoyproxy.io/docs/envoy/v1.22.0/configuration/observability/access_log/usage#command-operators) to extract and format data about a `TCP` connection or an `HTTP` request.

> E.g.,

Reviewer 3: ```suggestion
For example:
```

```
%START_TIME% %KUMA_SOURCE_SERVICE% => %KUMA_DESTINATION_SERVICE% %DURATION%
```

> where `%START_TIME%` and `%KUMA_SOURCE_SERVICE%` are examples of available _command operators_.

Reviewer 3: ```suggestion
`%START_TIME%` and `%KUMA_SOURCE_SERVICE%` are examples of available _command operators_.
```

All _command operators_ [defined by Envoy](https://www.envoyproxy.io/docs/envoy/v1.22.0/configuration/observability/access_log/usage#command-operators) are supported, along with additional _command operators_ defined by Kuma:

| Command Operator                     | Description                                                      |
|--------------------------------------|------------------------------------------------------------------|

> | `%KUMA_MESH%`                        | name of the mesh in which traffic is flowing                     |
> | `%KUMA_SOURCE_SERVICE%`              | name of a `service` that is the `source` of traffic              |
> | `%KUMA_DESTINATION_SERVICE%`         | name of a `service` that is the `destination` of traffic         |
> | `%KUMA_SOURCE_ADDRESS_WITHOUT_PORT%` | address of a `Dataplane` that is the `source` of traffic         |
> | `%KUMA_TRAFFIC_DIRECTION%`           | direction of the traffic, `INBOUND`, `OUTBOUND` or `UNSPECIFIED` |

Reviewer 3: ```suggestion
| `%KUMA_MESH%`                        | Name of the mesh in which traffic is flowing.                     |
| `%KUMA_SOURCE_SERVICE%`              | Name of a `service` that is the `source` of traffic.              |
| `%KUMA_DESTINATION_SERVICE%`         | Name of a `service` that is the `destination` of traffic.         |
| `%KUMA_SOURCE_ADDRESS_WITHOUT_PORT%` | Address of a `Dataplane` that is the `source` of traffic.         |
| `%KUMA_TRAFFIC_DIRECTION%`           | Direction of the traffic, `INBOUND`, `OUTBOUND`, or `UNSPECIFIED`. |
```

---
title: MeshAccessLog (beta)
---

With the MeshAccessLog policy you can easily set up access logs on every data plane in a mesh.

{% warning %}
This policy uses new policy matching algorithm and is in beta state,
it should not be mixed with [TrafficLog](/docs/{{ page.version }}/policies/trafficlog).
{% endwarning %}

{% tip %}
In the rest of this page we assume you have already configured your observability tools to work with Kuma.
If you haven't already read the [observability docs](/docs/{{ page.version }}/explore/observability).
{% endtip %}

## TargetRef support matrix

| TargetRef type    | top level | to  | from |
|-------------------|-----------|-----|------|
| Mesh              | ✅         | ✅   | ✅    |
| MeshSubset        | ✅         | ❌   | ❌    |
| MeshService       | ✅         | ✅   | ❌    |
| MeshServiceSubset | ✅         | ❌   | ❌    |

> | MeshGatewayRoute  | ✅         | ❌   | ❌    |
> | MeshHTTPRoute     | ❌         | ❌   | ❌    |

Reviewer: ```suggestion
```

These aren't actually supported/don't exist

Author: I was basing them on validator code so we should remove that from validation as well. https://github.com/kumahq/kuma/blob/b6fdbda848230bb26925f76c94fb6e7248403b43/pkg/plugins/policies/meshaccesslog/api/v1alpha1/validator.go#L32

    - key: "kuma_mesh"
      value: "%KUMA_MESH%"
    - key: 'method'
      value: '"%REQ(:METHOD)%'
    - key: "path"
      value: "%REQ(X-ENVOY-ORIGINAL-PATH?:PATH)%"
    - key: 'protocol'
      value: '%PROTOCOL%'
    - key: "response_code"
      value: "%RESPONSE_CODE%"
    - key: "response_flags"
      value: "%RESPONSE_FLAGS%"
    - key: "bytes_received"
      value: "%BYTES_RECEIVED%"
    - key: "bytes_sent"
      value: "%BYTES_SENT%"
    - key: "duration_ms"
      value: "%DURATION%"
    - key: "upstream_service_time"
      value: "%RESP(X-ENVOY-UPSTREAM-SERVICE-TIME)%"
    - key: 'x_forwarded_for'
      value: '"%REQ(X-FORWARDED-FOR)%"'
    - key: 'user_agent'
      value: '"%REQ(USER-AGENT)%"'
    - key: 'request_id'
      value: '"%REQ(X-REQUEST-ID)%"'
    - key: 'authority'
      value: '"%REQ(:AUTHORITY)%"'
    - key: "kuma_source_service"
      value: "%KUMA_SOURCE_SERVICE%"
    - key: "kuma_destination_service"
      value: "%KUMA_DESTINATION_SERVICE%"
    - key: "kuma_source_address_without_port"
      value: "%KUMA_SOURCE_ADDRESS_WITHOUT_PORT%"
    - key: "upstream_host"
      value: "%UPSTREAM_HOST%"
```

</details>

### Backends

#### TCP

> You can configure a `TCP` backend with an address:

Reviewer: ```suggestion
You can configure a TCP backend with an address:
```

```yaml
backends:
  - tcp:
      address: 127.0.0.1:5000
```

#### File

> You can configure a `File` backend with a path:

Reviewer: ```suggestion
You can configure a file backend with a path:
```

Can we be deliberate with our formatting here? Why "a `File` backend"? `File` makes me think of a resource called File.

```yaml
backends:
  - file:
      path: /tmp/access.log
```

## Examples

{% tabs meshaccesslog useUrlFragment=false %}
{% tab meshaccesslog Kubernetes %}

Full example:
```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshAccessLog
metadata:
  name: default
  namespace: kuma-system
  labels:
    kuma.io/mesh: default # optional, defaults to `default` if unset
spec:
  targetRef:
    kind: Mesh
  default:
    backends:
      - tcp:
          address: 127.0.0.1:5000
          format:
            json:
              - key: "start_time"
                value: "%START_TIME%"
      - file:
          path: /tmp/access.log
          format:
            plain: '[%START_TIME%]'
```

Apply the configuration with `kubectl apply -f [..]`.

{% endtab %}

> {% tab meshaccesslog Universal %}

Reviewer: Is it really useful to have Universal and Kubernetes examples _everywhere_? Very noisy IMO and a burden to maintain.

Author: Since every example on the whole docs has this we should discuss it with the whole team (moving this to slack).

Full example:
```yaml
type: MeshAccessLog
name: default
mesh: default
spec:
  targetRef:
    kind: Mesh
  default:
    backends:
      - tcp:
          address: 127.0.0.1:5000
          format:
            json:
              - key: "start_time"
                value: "%START_TIME%"
      - file:
          path: /tmp/access.log
          format:
            plain: '[%START_TIME%]'
```

> Apply the configuration with `kumactl apply -f [..]` or with the [HTTP API](../../reference/http-api).

Reviewer: More noisy than helpful IMO

{% endtab %}
{% endtabs %}

## Logging external services

When running Kuma on Kubernetes you can also log the traffic to external services.

> To do it use `MeshService` as a targetRef target.

Reviewer: I think we can always use `targetRef` in a mono font.

> To do it use `MeshService` as a targetRef target.

Reviewer: If this is true we should have an e2e test. I was under the impression this will go over the pass_through cluster and thus only loggable using `kind: Mesh`

Author: I wasn't sure so I asked on slack.

|--------------------------------------|------------------------------------------------------------------|
| `%KUMA_MESH%`                        | name of the mesh in which traffic is flowing                     |
| `%KUMA_SOURCE_SERVICE%`              | name of a `service` that is the `source` of traffic              |
| `%KUMA_DESTINATION_SERVICE%`         | name of a `service` that is the `destination` of traffic         |
| `%KUMA_SOURCE_ADDRESS_WITHOUT_PORT%` | address of a `Dataplane` that is the `source` of traffic         |
| `%KUMA_TRAFFIC_DIRECTION%`           | direction of the traffic, `INBOUND`, `OUTBOUND` or `UNSPECIFIED` |

All additional access log _command operators_ are valid to use with both `TCP` and `HTTP` traffic.

If a _command operator_ is specific to `HTTP` traffic, such as `%REQ(X?Y):Z%` or `%RESP(X?Y):Z%`, it will be replaced by a symbol "`-`" in case of `TCP` traffic.

Internally, Kuma [determines traffic protocol](/docs/{{ page.version }}/policies/protocol-support-in-kuma) based on the value of `kuma.io/protocol` tag on the `inbound` interface of a `destination` `Dataplane`.

There are two types of `format` - `plain` and `json`.

> Plain will accept a string with _command operators_ and produce a string output.

Reviewer 3: ```suggestion
Plain accepts a string with _command operators_ and produces a string output.
```

          - file:
              path: /tmp/access.log
              format:
                plain: '[%START_TIME%]'
```

Apply the configuration with `kubectl apply -f [..]`.

{% endtab %}
{% tab meshaccesslog-multiple-backends Universal %}

```yaml
type: MeshAccessLog
name: default
mesh: default
spec:
  targetRef:
    kind: Mesh
  from:
    - targetRef:
        kind: Mesh
      default:
        backends:
          - file:
              address: 127.0.0.1:5000
              format:
                json:
                  - key: "start_time"
                    value: "%START_TIME%"
          - file:
              path: /tmp/access.log
              format:
                plain: '[%START_TIME%]'
```

Apply the configuration with `kumactl apply -f [..]` or with the [HTTP API](../../reference/http-api).

{% endtab %}
{% endtabs %}

## Log all incoming and outgoing traffic

{% tabs meshaccesslog-all useUrlFragment=false %}
{% tab meshaccesslog-all Kubernetes %}

```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshAccessLog
metadata:
  name: default
  namespace: kuma-system
  labels:
    kuma.io/mesh: default # optional, defaults to `default` if unset
spec:
  targetRef:
    kind: Mesh

>   from: # delete this section if you do not want to log incoming traffic

Reviewer 3: ```suggestion
  from: # delete this section if you don't want to log incoming traffic
```

    - targetRef:
        kind: Mesh
      default:
        backends:
          - file:
              path: /tmp/access.log

>   to: # delete this section if you do not want to log outgoing traffic

Reviewer 3: ```suggestion
  to: # delete this section if you don't want to log outgoing traffic
```

    - targetRef:
        kind: Mesh
      default:
        backends:
          - file:
              path: /tmp/access.log
```

Apply the configuration with `kubectl apply -f [..]`.

{% endtab %}
{% tab meshaccesslog-all Universal %}

```yaml
type: MeshAccessLog
name: default
mesh: default
spec:
  targetRef:
    kind: Mesh

>   from: # delete this section if you do not want to log incoming traffic

Reviewer 3: ```suggestion
  from: # delete this section if you don't want to log incoming traffic
```

    - targetRef:
        kind: Mesh
      default:
        backends:
          - file:
              path: /tmp/access.log

>   to: # delete this section if you do not want to log outgoing traffic

Reviewer 3: ```suggestion
  to: # delete this section if you don't want to log outgoing traffic
```

    - targetRef:
        kind: Mesh
      default:
        backends:
          - file:
              path: /tmp/access.log
```

Apply the configuration with `kumactl apply -f [..]` or with the [HTTP API](../../reference/http-api).

{% endtab %}
{% endtabs %}

## Logging external services

> To do it use `MeshService` as a `targetRef` target and match external service `name`.

Reviewer 3: ```suggestion
To log external services, use `MeshService` as a `targetRef` target and match external service `name`.
```

> ## Builtin gateway

Reviewer 3: ```suggestion
## Select a built-in gateway
```

> You can select a builtin gateway using the `kuma.io/service` value. A current limitation is that traffic routed from a gateway to a service will be logged by that gateway as having destination `"*"`.

Reviewer 3: ```suggestion
You can select a built-in gateway using the `kuma.io/service` value. A current limitation is that traffic routed from a gateway to a service is logged by that gateway as having destination `"*"`.
```

---
title: MeshAccessLog (beta)
---

With the MeshAccessLog policy you can easily set up access logs on every data plane proxy in a mesh.

{% warning %}
This policy uses new policy matching algorithm and is in beta state,
it should not be mixed with [TrafficLog](/docs/{{ page.version }}/policies/traffic-log).
{% endwarning %}

{% tip %}
In the rest of this page we assume you have already configured your observability tools to work with Kuma.
If you haven't already read the [observability docs](/docs/{{ page.version }}/explore/observability).
{% endtip %}

> ## TargetRef support matrix
> 
> | TargetRef kind    | top level | to  | from |
> |-------------------|-----------|-----|------|
> | Mesh              | ✅         | ✅   | ✅    |
> | MeshSubset        | ✅         | ❌   | ❌    |
> | MeshService       | ✅         | ✅   | ❌    |
> | MeshServiceSubset | ✅         | ❌   | ❌    |

Reviewer: ```suggestion
## `targetRef` support matrix

| `targetRef.kind`    | top level | to  | from |
| ------------------- | --------- | --- | ---- |
| `Mesh`              | ✅        | ✅  | ❌   |
| `MeshSubset`        | ✅        | ❌  | ❌   |
| `MeshService`       | ✅        | ❌  | ✅   |
| `MeshServiceSubset` | ✅        | ❌  | ❌   |
```

Until we have our helper :wink:

spec:
  targetRef:
    kind: Mesh
  from: # delete this section if you do not want to log incoming traffic
    - targetRef:
        kind: Mesh
      default:
        backends:
          - file:
              path: /tmp/access.log
  to: # delete this section if you do not want to log outgoing traffic
    - targetRef:
        kind: Mesh
      default:
        backends:
          - file:
              path: /tmp/access.log
```

Apply the configuration with `kubectl apply -f [..]`.

{% endtab %}
{% tab meshaccesslog-all Universal %}

```yaml
type: MeshAccessLog
name: default
mesh: default
spec:
  targetRef:
    kind: Mesh
  from: # delete this section if you do not want to log incoming traffic
    - targetRef:
        kind: Mesh
      default:
        backends:
          - file:
              path: /tmp/access.log
  to: # delete this section if you do not want to log outgoing traffic
    - targetRef:
        kind: Mesh
      default:
        backends:
          - file:
              path: /tmp/access.log
```

Apply the configuration with `kumactl apply -f [..]` or with the [HTTP API](../../reference/http-api).

{% endtab %}
{% endtabs %}

## Logging traffic going outside the Mesh

> To target [External Service](/docs/{{ page.version }}/policies/external-services#usage) use `MeshService` as a `targetRef` target and match external service `name`.
> 
> To target traffic going through [passthrough mode](/docs/{{ page.version }}/policies/external-services) use `Mesh` as a `targerRef`. This will cause case `%KUMA_DESTINATION_SERVICE%` will have value `external` and `%UPSTREAM_HOST%` will have an IP of the service.

Reviewer: ```suggestion
To target [`ExternalServices`](/docs/{{ page.version }}/policies/external-services#usage), use `MeshService` as the `targetRef` kind with `name` set to  
the `kuma.io/service` value.

To target other non-mesh traffic, i.e. [passthrough traffic](/docs/{{ page.version }}/policies/mesh/#controlling-the-passthrough-mode), use `Mesh` as the `targetRef` kind. In this case, `%KUMA_DESTINATION_SERVICE%` is set to `external` and `%UPSTREAM_HOST%` is set to IP of the service.
```

Author: I'll apply this but I don't know why `UPSTREAM_HOST` is no longer relevant here.

Reviewer: I don't know why it was ever relevant, it will have the obvious value like every other Envoy command operator.
